### DISCUSSION TITLE
Unicode selection for custom icons

### DISCUSSION THREAD

**Discussion by moving-bits** [cgeo/cgeo#9717] [Q&A] [answered]
On current master we have added an option to select custom cache icons which are based on the regular Unicode fonts. Currently the offered characters are just a bunch of randomly chosen characters (50 consecutive characters starting at a certain position), but I would like to discuss which characters may be useful, and how many.

IMHO if we select app. 50 characters, we can select them from a single mask, but if we decide to select more (which would be ok as well), we should group them in some clearly distinguishable categories.

It's pretty hard to find a good selection - really depends on what people would like to markup, especially if it's not only about geocaching. - So I tried to focus on geocaching first and some other free-time activities, and prefer generic symbols over very specific ones.

I scanned the Unicode character list (found here: https://unicode.org/emoji/charts-12.0/full-emoji-list.html), which is neatly organized by emoji groups, which I use as reference in the table below.

While selecting groups I came to the conclusion that offering a few categories will help our users in selecting the icons. I came up with the following categories to start with:
- people
- activities 
- food
- places
- symbols

I propose to select icons from the following emoji groups:

|c:geo category|emoji groups (from Unicode list)|comment|
|---|---|---|
|people|face-smiling, face-affection, emotion, person, face-costume, plant-other, |select a couple from each group|
|food|food-*, drink |select a couple from each group|
|activity|person-activity, person-sport, sport|maybe haircut, walking, manual wheelchair, running|
|activity|person-sport, sport, game, arts & crafts|select a couple from each group|
|places|place-map|globes|
|places|place-geographic, place-building, transport-*|select a couple from each group|
|symbols|keycap|0-9|
|symbols|geometric|a selection of colored dots and squares|
|symbols|flag|some generic flags|

I would (mostly) skip the following emoji groups: 
face-tongue, face-hand, face-neutral-skeptical, face-sleepy, face-unwell, face-hat, face-glasses, face-concerned, face-negative, cat-face, monkey-face, people & body, hand-fingers-partial, hand-single-finger, hand-fingers-closed, hands, hand-prop, body-parts, person-gesture, person-role, person-fantasy, person-resting, family, person-symbol, component, animals-mammal, animal-bird, animal-amphibian, animal-reptile, animal-marine, animal-bug, plant-flower, dishware, place-religious, place-other, hotel, time, sky & weather, event, clothing, soumd, music, musical-instrument, phone, computer, light & video, book-paper, money, mail, writing, office, lock, tool, science, medical, household, other-object, warning, arrow, religion, zodiac, av-symbol, gender, other-symbol, alphanum, country-flag, subdivision-flag

Maybe we can scan again for some (more or less) typical geocaching equipment needed for special caches, eg. a ladder, boots etc.

What do you think?

I you generaly agree to the above pre-selection, I would start selecting specific symbols from each mentioned group.

**Comment by fm-sys** [comment]
I also had a look at the emoji list. The Only problem is, that most of these aren't supported at Android 5, our minimum API level. 

So we either have to choose the emojis very carefully (many useful emojis wouldn't be supported) or we need to use the emoji compat support libary:

https://developer.android.com/guide/topics/ui/look-and-feel/emoji-compat

**Reply by moving-bits** [reply]
Is there some chart which API version supports which emoji?

**Reply by moving-bits** [reply]
Found some info on https://developer.android.com/guide/topics/resources/internationalization:

Platform (API level) | ICU | CLDR | Unicode
-- | -- | -- | --
Android 5.0 (API levels 21–22) | 53 | 25 | 6.3
Android 6.0 (API level 23) | 55.1 | 27.0.1 | 7.0
Android 7.0 - 7.1 (API levels 24 - 25) | 56 | 28 | 8.0
Android 8.0 - 8.1 (API levels 26 - 27) | 58.2 | 30.0.3 | 9.0
Android 9 (API level 28) | 60.2 | 32.0.1 | 10.0
Android 10 (API level 29) | 63.2 | 34 | 11.0

Looks like we need to strip down the selection quite a bit (or include compat lib - need to dive into that a bit deeper).

**Reply by moving-bits** [reply]
Did a (very quick) scan for v5 Unicode (the last before v6.3 I could find):
Even that list looks pretty useful: https://unicode.org/emoji/charts-5.0/full-emoji-list.html (although the HTML file "only" has 35 MB instead of the > 50 MB for v12.0 ;-))

**Reply by fm-sys** [reply]
Here is the full emoji list for Android 5 https://emojipedia.org/google/android-5.0/

Regarding the topic of geocaching we would be unfortunately quite limited, so I would clearly vote for using the compat lib. In the docs it sounds like the emojis aren't stored in the APK but can be downloaded during run time...

**Reply by fm-sys** [reply]
OK, I took another look, and you're right that it might be sufficient without compat lib. But we should include the colored dots and squares for higher API's too...

**Comment by fm-sys** [comment]
I would only use the 🧗 (climbing) of `person` section and for `person-sport` we should add emojis like boat, swimming, ski, etc.

I think we can skip `face-costume`, and mostly `food & drink` ...
...but add all four emojis from `person-symbol` as they seam to be quite useful

Oh, and question/exclamation mark must not be missing (⁉, ❓, ❗)


But however: Already a great preselection!

**Comment by Lineflyer** [comment]
My few cents:
- Selecting meaningful emojis, which users can use to describe the "attributes" of a cache (like the mentioned climbing, swimming) are far more meaningful than just a bunch of random emojis (like fruits). So it makes sense to offer others and have categories
- I would however reduce to some potentially descriptive emojis on some categories (maybe driven or limited by the minimum supported emojis according to minSDK)
- I would not suggest to take the effort to introduce just another compat library for such "less important" feature. Better to evolve with more emojis as soon as we might raise the minSDK in future.

**Comment by moving-bits** [comment, marked as answer]
Reminder to myself: Check `Canvas.hasGlyph(String)` to check whether the current systems support a specific Unicode character, and skip it in the selection box, if not. (Caveat: Only supported on API 23+)

**Reply by moving-bits** [reply]
implemented with #9735 - works
